azkaban-common/src/main/java/azkaban/project/DirectoryFlowLoader.java

```diff
@@ -329,15 +329,18 @@ private void buildFlowsFromDependencies() {
         FlowLoaderUtils.addEmailPropsToFlow(flow, jobProp);
 
         flow.addAllFlowProperties(this.flowPropsList);
-        constructFlow(flow, base, visitedNodes);
+        Set<String> visitedNodesOnPath = new HashSet<>();
+        constructFlow(flow, base, visitedNodesOnPath, visitedNodesEver);
+
         flow.initialize();
         this.flowMap.put(base.getId(), flow);
       }
     }
   }
 
-  private void constructFlow(final Flow flow, final Node node, final Set<String> visited) {
-    visited.add(node.getId());
+  private void constructFlow(final Flow flow, final Node node, final Set<String> visitedOnPath, final Set<String> visitedEver) {
+    visitedOnPath.add(node.getId());
+    visitedEver.add(flow.getId() + ":" + node.getId());
```

Comment:
Originally, I thought it would be unique as well.  However, your unit tests quickly disproved that theory!  It was happening when 2 different flows had jobs with the same name.  I believe the conflict in the tests was `innerJobA` which appeared in multiple flows. Although, this may imply that `visitedEver` is simply not being reset between flows or something.

Reply:
I see. For embedded flows, they could share the same inner job. I guess if you move visitedNodesEver inside the for loop, basically the same place as visitedNodesOnPath, then you probably don't need to use "flowId:nodeId" to distinguish the nodes.

Reply:
I'll take a look at that.  Thanks!

Reply:
If I understand you correctly, moving the declaration of `visitedNodesEver` into the for loop would result in the contents of `visitedNodesEver` being identical to `visitedNodesOnPath` which would defeat the purpose.  I think we need to keep the declaration where it is and simple fully qualify the node id with the flow id (as it is now) in order to safely ensure that we only visit a node once

Reply:
My bad.  I took another look at it, and your suggestion will work.